LibraBFT/Impl/Consensus/RoundManager/Properties.agda

```diff
@@ -123,69 +141,155 @@ module processProposalMSpec (proposal : Block) where
+    constructor mkContract
+    field
+      noEpochChange : NoEpochChange pre post
+      noVoteOuts    : NoVoteOuts outs
+      noVote        : NoVoteCorrect pre post true
+
+  postulate
+    contract'
+      : ∀ pre → LBFT-weakestPre (ensureRoundAndSyncUpM now messageRound syncInfo author helpRemote) (Contract pre) pre
+
+  contract : ∀ pre Post → RWST-Post-⇒ (Contract pre) Post → LBFT-weakestPre (ensureRoundAndSyncUpM now messageRound syncInfo author helpRemote) Post pre
+  contract pre Post pf =
+    LBFT-⇒ (Contract pre) Post pf (ensureRoundAndSyncUpM now messageRound syncInfo author helpRemote) pre (contract' pre)
+
+module processProposalMsgMSpec
+  (now : Instant) (pm : ProposalMsg) where
+
+  open processProposalMsgM now pm
+
+  epoch = pm ^∙ pmProposal ∙ bEpoch
+  round = pm ^∙ pmProposal ∙ bRound
+
+  record Contract (pre : RoundManager) (_ : Unit) (post : RoundManager) (outs : List Output) : Set where
+    constructor mkContract
+    field
+      noEpochChange : NoEpochChange pre post
+      outsCorrect   : NoVote⊎VoteMsgOutsCorrect pre post outs epoch round
+
+  contract' : ∀ pre → LBFT-weakestPre (processProposalMsgM now pm) (Contract pre) pre
+  contract' pre rewrite processProposalMsgM≡ = contract
+    where
+    contractBail : ∀ outs → NoVoteOuts outs → Contract pre unit pre outs
+    contractBail outs nvo =
+      mkContract reflNoEpochChange (Left (true , (mkNoVoteMsgOutsCorrect nvo (Left reflNoVoteCorrect))))
+
+    contract : LBFT-weakestPre step₀ (Contract pre) pre
+    proj₁ contract ≡nothing = contractBail _ refl
+    proj₂ contract = contract-step₁
+      where
+      module _ (pAuthor : Author) (pAuthor≡ : pm ^∙ pmProposer ≡ just pAuthor) where
+        step₂-pf : RWST-Post-⇒ _ (RWST-weakestPre-bindPost unit step₂ (Contract pre))
+
+        contract-step₁ : LBFT-weakestPre (step₁ pAuthor) (Contract pre) pre
+        contract-step₁ =
+          ensureRoundAndSyncUpMSpec.contract now round (pm ^∙ pmSyncInfo) pAuthor true pre
+            (RWST-weakestPre-bindPost unit step₂ (Contract pre)) step₂-pf
+
+        step₂-pf (Left      e) st outs (ensureRoundAndSyncUpMSpec.mkContract noEpochChange noVoteOuts noVote) ._ refl =
+          mkContract noEpochChange (Left (true , (mkNoVoteMsgOutsCorrect (++-NoVoteOuts outs _ noVoteOuts refl) (Left noVote))))
+        step₂-pf (Right false) st outs (ensureRoundAndSyncUpMSpec.mkContract noEpochChange noVoteOuts noVote) ._ refl ._ refl =
+          mkContract noEpochChange (Left (true , (mkNoVoteMsgOutsCorrect (++-NoVoteOuts outs _ noVoteOuts refl) (Left noVote))))
+        step₂-pf (Right  true) st outs (ensureRoundAndSyncUpMSpec.mkContract noEpochChange noVoteOuts noVote) ._ refl =
+          processProposalMSpec.contract (pm ^∙ pmProposal) st (RWST-Post++ (Contract pre) outs) step₃-pf
+          where
+          step₃-pf : RWST-Post-⇒ _ (RWST-Post++ (Contract pre) outs)
```

Comment:
> Similar comment as above. Why not define/prove step₃-pf separately rather than in a where clause?

Same reason as above --- it's convenient here since `outs` is already in scope.